### PR TITLE
Port SqlParameter changes from netcore to netfx

### PR DESCRIPTION
**Pull request by Wraith2**
Ports the changes in SqlParameters perf optimizations to netfx https://github.com/dotnet/SqlClient/pull/209. 
This is a prerequisite for https://github.com/dotnet/SqlClient/issues/19

### CODE REVIEW

**Comment by JRahnama** [comment]
@Wraith2 Thanks for the PR. The test pipelines are setup, but there are some adjustments left to be done. I saw some failures on net462 and not sure if you were able to see them.
image

**Comment by Wraith2** [comment]
It turns out you have to actually set the parameters field so they can be sent to the server or all sorts of things stop working.

The CI doesn't seem happy because legs i haven't change (enclave ones) are failing for net6.0 builds. I'm also not sure which ones i should be paying attention to. Do we still have legacy checks?

**Comment by JRahnama** [comment]
> It turns out you have to actually set the parameters field so they can be sent to the server or all sorts of things stop working.
> 
> The CI doesn't seem happy because legs i haven't change (enclave ones) are failing for net6.0 builds. I'm also not sure which ones i should be paying attention to. Do we still have legacy checks?

I will take the old one out today. The one I was working on, this will be primary one today, is CI-SqlClient-To-Be-Main.

**Comment by Wraith2** [comment]
Most legs are passing but some fail with EXEC(0)

**Comment by Wraith2** [comment]
Can you kick the CI again please.

**Comment by Wraith2** [comment]
The failures are odd but appear repeatable. They're in Always Encrypted tests which i have no way to debug locally because of the sql server requirements. I'm going to need some help from the team to investigate why those failures are occuring.

As i mentioned in the first post this is a blocking issue for the sql batch api, without these changes there is no way to access the parameter collection for the set of rpc's in the batch.

**Comment by DavoudEshtehari** [comment]
The failures don't require Always Encrypted enabled.

**Comment by Wraith2** [comment]
Ok, all green now. quick merge it before I break it again!

File: src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlCommand.cs [hunk 1]
```diff
@@ -4573,20 +4564,23 @@ private void PrepareDescribeParameterEncryptionRequest(_SqlRPC originalRpcReques
...
+                Debug.Assert(originalRpcRequest.systemParamCount > 0,
                     "originalRpcRequest didn't have at-least 1 parameter in BatchRPCMode, in PrepareDescribeParameterEncryptionRequest.");
-                text = (string)originalRpcRequest.parameters[0].Value;
-                sqlParam = GetSqlParameterWithQueryText(text);
+                text = (string)originalRpcRequest.systemParams[0].Value;
+                //@tsql
+                SqlParameter tsqlParam = describeParameterEncryptionRequest.systemParams[0];
+                tsqlParam.SqlDbType = ((text.Length << 1) <= TdsEnums.TYPE_SIZE_LIMIT) ? SqlDbType.NVarChar : SqlDbType.NText;
+                tsqlParam.Value = text;
+                tsqlParam.Size = text.Length;
```
[the remark this answers is not in the document]
Comment (Wraith2) [reply src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlCommand.cs:4583]: Done.

**Review by DavoudEshtehari — APPROVED** [review]

**Review by JRahnama — APPROVED** [review]

**Review by lcheunglci — APPROVED** [review]

**Pull request merged**